### PR TITLE
[WIP] PIM-3832: Rework product converter

### PR DESCRIPTION
**Pull request by solivier**

### CODE REVIEW

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/ConvertToStructuredField.php [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+use Pim\Component\Connector\ArrayConverter\Flat\ProductAttributeFieldExtractor;
+
+/**
+ * Product Flat Converter
+ *
+ * @author    Olivier Soulet <olivier.soulet@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ConvertToStructuredField
```
Comment (nidup) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/ConvertToStructuredField.php]: StructuredFieldConvertor

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/ConvertToStructuredField.php [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+     *
+     * @param string $value          The value content
+     * @param array  $fieldNameInfos The field informations
+     *
+     * @return array
+     */
+    protected function formatValueData($value, $fieldNameInfos)
+    {
+        switch ($fieldNameInfos['attribute']->getAttributeType()) {
+            case 'pim_catalog_price_collection':
```
Comment (nidup) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/ConvertToStructuredField.php]: we should explode all this small transformation in dedicated tagged services, to allow to the integrator to declare their own for their attribute type. we can do something close to what we did for the setters for instance https://github.com/akeneo/pim-community-dev/tree/master/src/Pim/Bundle/CatalogBundle/Updater/Setter

File: src/Pim/Bundle/ConnectorBundle/Resources/config/array_converters.yml [hunk 3]
```diff
@@ -30,7 +51,91 @@ services:
...
+            - '@pim_catalog.repository.attribute'
+            - '@pim_catalog.resolver.attribute_values'
+
+
     pim_base_connector.array_converter.flat.association_field_resolver:
         class: %pim_base_connector.array_converter.flat.association_field_resolver.class%
         arguments:
             - '@pim_catalog.repository.association_type'
+
+    #tests
```
Comment (nidup) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/array_converters.yml]: remove this com

File: spec/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/AbstractConverterSpec.php [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+namespace spec\Pim\Component\Connector\ArrayConverter\Flat\Product\Converter;
+
+use PhpSpec\ObjectBehavior;
+use Pim\Component\Connector\ArrayConverter\Flat\Product\Splitter\FieldSplitter;
+
+class AbstractConverterSpec extends ObjectBehavior
```
Comment (nidup) [inline spec/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/AbstractConverterSpec.php]: remove this ?

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/ConverterInterface.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+interface ConverterInterface
+{
+    /**
+     * Converts value
+     *
+     * @param string $fieldNameInfo
+     * @param string $value
+     *
+     * @return array
+     *
```
Comment (nidup) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/ConverterInterface.php]: run php-cs-fixer

File: src/Pim/Bundle/ConnectorBundle/DependencyInjection/Compiler/RegisterConverterPass.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Register converters
+ *
+ * @author    Olivier Soulet <olivier.soulet@akeneo.com>
+ * @copyright 2014 Akeneo SAS (http://www.akeneo.com)
```
Comment (skeleton) [inline src/Pim/Bundle/ConnectorBundle/DependencyInjection/Compiler/RegisterConverterPass.php]: 2015

File: src/Pim/Component/Connector/ArrayConverter/Flat/ProductOptionalAttributeFieldExtractor.php [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+use Pim\Bundle\CatalogBundle\Repository\CurrencyRepositoryInterface;
+
+/**
+ * Extracts attribute field information
+ *
+ * @author    Olivier Soulet <olivier.soulet@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ProductOptionalAttributeFieldExtractor
```
Comment (nidup) [inline src/Pim/Component/Connector/ArrayConverter/Flat/ProductOptionalAttributeFieldExtractor.php]: Looks more like a Resolver -> ProductOptionalAttributeFieldResolver

File: src/Pim/Component/Connector/ArrayConverter/Flat/ProductOptionalAttributeFieldExtractor.php [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    ) {
+        $this->currencyRepository  = $currencyRepository;
+        $this->attributeRepository = $attributeRepository;
+        $this->valuesResolver      = $valuesResolver;
+    }
+
+    /**
+     * @return array
+     */
+    public function getOptionalAttributeFields()
```
Comment (nidup) [inline src/Pim/Component/Connector/ArrayConverter/Flat/ProductOptionalAttributeFieldExtractor.php]: resolveOptionalAttributeFields

File: src/Pim/Component/Connector/ArrayConverter/Flat/ProductToStandardConverter.php [hunk 9]
```diff
@@ -186,126 +176,27 @@ protected function convertToStructuredField($column, $value)
...
-     *
-     * @param string $value          The value content
-     * @param array  $fieldNameInfos The field informations
+     * @param string $column
+     * @param string $value
      *
      * @return array
      */
-    protected function formatPrice($value, $fieldNameInfos)
+    public function formatValue($column, $value)
```
Comment (nidup) [inline src/Pim/Component/Connector/ArrayConverter/Flat/ProductToStandardConverter.php]: rename to convertValue

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/BooleanConverter.php [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+class BooleanConverter extends AbstractConverter
+{
+    /**
+     * @param FieldSplitter $fieldSplitter
+     * @param array         $supportedFieldType
+     */
+    public function __construct(
+        FieldSplitter $fieldSplitter,
+        array $supportedFieldType
+    ) {
```
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/BooleanConverter.php]: one line

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/BooleanConverter.php [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        if ('' === $value) {
+            return null;
+        }
+
+        $data = (bool) $value;
+
+        return [$fieldNameInfo['attribute']->getCode() => [[
+            'locale' => $fieldNameInfo['locale_code'],
+            'scope'  => $fieldNameInfo['scope_code'],
+            'data'   => $data,
```
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/BooleanConverter.php]: Are you sure key of $data is really "data" ? not "value" ?

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/ConverterRegistry.php [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    public function getConverter($field)
+    {
+        foreach ($this->converters as $converter) {
+            if ($converter->supportsField($field)) {
+                return $converter;
+            }
+        }
+
+        return null;
```
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/ConverterRegistry.php]: Maybe we should throw a \LogicException if there is no converter ?

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/IdentifierConverter.php [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+class IdentifierConverter extends AbstractConverter
+{
+    /**
+     * @param FieldSplitter $fieldSplitter
+     * @param array         $supportedFieldType
+     */
+    public function __construct(
+        FieldSplitter $fieldSplitter,
+        array $supportedFieldType
+    ) {
```
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/IdentifierConverter.php]: one line

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/MetricConverter.php [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        }
+
+        //Due to the multi column format for metrics
+        if (isset($fieldNameInfo['metric_unit'])) {
+            $unit = $fieldNameInfo['metric_unit'];
+        } else {
+            list($value, $unit) = $this->fieldSplitter->splitUnitValue($value);
+        }
+
+
```
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/MetricConverter.php]: one blank line

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/PriceConverter.php [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+        return [$fieldNameInfo['attribute']->getCode() => [[
+            'locale' => $fieldNameInfo['locale_code'],
+            'scope'  => $fieldNameInfo['scope_code'],
+            'data'   => $data,
+        ]]];
+    }
+
+    /**
+     * @param string $priceValue
+     * @param array  $fieldNameInfo
```
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/PriceConverter.php]: bad phpdoc

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/ConverterInterface.php [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+namespace Pim\Component\Connector\ArrayConverter\Flat\Product\Converter;
+
+/**
+ * Converts data
+ *
+ * @author    Olivier Soulet <olivier.soulet@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface ConverterInterface
```
Comment (nidup) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/ConverterInterface.php]: still thinking to the naming of this one, confusing with the StandardArrayConverterInterface, as this set of converters are dedicated to the product array convertion, I would maybe make them more "specific".

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/OptionsResolverConverter.php [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    protected $attributeRepository;
+
+    /** @var CurrencyRepositoryInterface */
+    protected $currencyRepository;
+
+    /** @var AttributeValuesResolver */
+    protected $valuesResolver;
+
+    /** @var array */
+    protected $optionalAttributeFields;
```
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/OptionsResolverConverter.php]: not used

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/OptionsResolverConverter.php [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+class OptionsResolverConverter
+{
+    /** @var AttributeRepositoryInterface */
+    protected $attributeRepository;
+
+    /** @var CurrencyRepositoryInterface */
+    protected $currencyRepository;
+
+    /** @var AttributeValuesResolver */
+    protected $valuesResolver;
```
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/OptionsResolverConverter.php]: not used

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/OptionsResolverConverter.php [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class OptionsResolverConverter
+{
+    /** @var AttributeRepositoryInterface */
+    protected $attributeRepository;
+
+    /** @var CurrencyRepositoryInterface */
+    protected $currencyRepository;
```
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/OptionsResolverConverter.php]: not used

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/OptionsResolverConverter.php [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+ * Resolve product saving options for converter
+ *
+ * @author    Olivier Soulet <olivier.soulet@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class OptionsResolverConverter
+{
+    /** @var AttributeRepositoryInterface */
+    protected $attributeRepository;
```
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/OptionsResolverConverter.php]: not used

File: src/Pim/Component/Connector/ArrayConverter/Flat/ProductOptionalAttributeFieldExtractor.php [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        if (empty($this->optionalAttributeFields)) {
+            $attributes = $this->attributeRepository->findAll();
+            $currencyCodes = $this->currencyRepository->getActivatedCurrencyCodes();
+            $values = $this->valuesResolver->resolveEligibleValues($attributes);
+            foreach ($values as $value) {
+                if ($value['locale'] !== null && $value['scope'] !== null) {
```
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/Flat/ProductOptionalAttributeFieldExtractor.php]: yoda style please

File: src/Pim/Component/Connector/ArrayConverter/Flat/ProductOptionalAttributeFieldExtractor.php [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ProductOptionalAttributeFieldExtractor
+{
+    /** @var AttributeRepositoryInterface */
+    protected $attributeRepository;
+
+    /** @var ChannelRepositoryInterface */
+    protected $channelRepository;
```
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/Flat/ProductOptionalAttributeFieldExtractor.php]: not used

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/ConverterRegistry.php (hunk 12 above)
Comment (nidup) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/ConverterRegistry.php]: yes

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/BooleanConverter.php (hunk 11 above)
Comment (nidup) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/Converter/BooleanConverter.php]: we need to double check Setter use data but normalizer use value ... :(

**Pull request merged**